In [1]:
#import KNeighborsClassifier
from sklearn.neighbors import KNeighborsClassifier

In [2]:
#import numpy and pandas libraries 
import numpy as np
import pandas as pd

In [3]:
#read iris flower data set and assign it to pandas dataframe
df_iris = pd.read_csv('https://raw.githubusercontent.com/mpourhoma/CS4661/master/iris.csv')

In [4]:
#create feature matrix for iris dataset:
#this creates a list of feature names in the dataset
feature_cols = ['sepal_length', 'sepal_width', 'petal_length', 'petal_width']
#select features from the DataFrame
X = df_iris[feature_cols]

#printing list of features for checking purposes
X[::10]

sepal_length  sepal_width  petal_length  petal_width
0             5.1          3.5           1.4          0.2
10            5.4          3.7           1.5          0.2
20            5.4          3.4           1.7          0.2
30            4.8          3.1           1.6          0.2
40            5.0          3.5           1.3          0.3
50            7.0          3.2           4.7          1.4
60            5.0          2.0           3.5          1.0
70            5.9          3.2           4.8          1.8
80            5.5          2.4           3.8          1.1
90            5.5          2.6           4.4          1.2
100           6.3          3.3           6.0          2.5
110           6.5          3.2           5.1          2.0
120           6.9          3.2           5.7          2.3
130           7.4          2.8           6.1          1.9
140           6.7          3.1           5.6          2.4

In [5]:
#selects labels from DataFrame and set in y variable
y = df_iris['species']

#print list of labels for checking purposes
y[::10]

0          setosa
10         setosa
20         setosa
30         setosa
40         setosa
50     versicolor
60     versicolor
70     versicolor
80     versicolor
90     versicolor
100     virginica
110     virginica
120     virginica
130     virginica
140     virginica
Name: species, dtype: object

In [6]:
#import train_test_split library
#split data into training (60%) and testing set (40%)
from sklearn.model_selection import train_test_split

X_train, X_test, y_train, y_test = train_test_split(X, y, test_size = 0.4,random_state = 6)

In [7]:
#Instantiate object of KNeighborsClassifier with k = 3
k = 3
knn_kevin = KNeighborsClassifier(n_neighbors = k)

In [8]:
#train the model on training set
knn_kevin.fit(X_train, y_train)

KNeighborsClassifier(n_neighbors=3)

In [9]:
#test model on testing set
y_predict = knn_kevin.predict(X_test)
print(y_predict)

['setosa' 'virginica' 'setosa' 'setosa' 'virginica' 'versicolor'
 'virginica' 'setosa' 'virginica' 'versicolor' 'virginica' 'versicolor'
 'virginica' 'virginica' 'versicolor' 'versicolor' 'virginica'
 'versicolor' 'versicolor' 'setosa' 'setosa' 'virginica' 'setosa' 'setosa'
 'versicolor' 'versicolor' 'versicolor' 'virginica' 'setosa' 'versicolor'
 'setosa' 'versicolor' 'setosa' 'setosa' 'versicolor' 'virginica'
 'versicolor' 'virginica' 'versicolor' 'setosa' 'setosa' 'virginica'
 'versicolor' 'versicolor' 'setosa' 'setosa' 'versicolor' 'setosa'
 'setosa' 'versicolor' 'virginica' 'virginica' 'virginica' 'setosa'
 'virginica' 'setosa' 'setosa' 'setosa' 'versicolor' 'virginica']


In [10]:
#import accuracy score and find accuracy percentage
from sklearn.metrics import accuracy_score
accuracy = accuracy_score(y_test,y_predict)
print(accuracy)

0.9666666666666667


In [11]:
#save K values into a list
list = [1, 5, 7, 11, 15, 27, 59]
#create empty list to save accuracy values in
acc = []
#create forloop to iterate through list and find accuracy based on "K" values in list

for l in list:
    #set n_neighbors to values in list
    knn_kevin = KNeighborsClassifier(n_neighbors=l)
    #train data on training set
    knn_kevin.fit(X_train, y_train)
    #test data on testing set
    y_predict = knn_kevin.predict(X_test)
    #find accuracy of testing set vs actual lavels
    accuracy = accuracy_score(y_test, y_predict)
    #save the accuracy values into empty list "acc"
    acc.append(accuracy)
    
print(acc)
    

[0.95, 0.9833333333333333, 0.9666666666666667, 0.9666666666666667, 0.9333333333333333, 0.9166666666666666, 0.8166666666666667]


Does the accuracy always get better by increasing the value of K?

After k = 5 the accuracy of the predictions decreases so when K increases greatly the accuracy percentage decreases. 

In [12]:
#Creating new feature columns based on each of 4 features
#sepal_length
feature_col1 = ['sepal_length']
X1 = df_iris[feature_col1]
#sepal_width
feature_col2 = ['sepal_width']
X2 = df_iris[feature_col2]
#petal_length
feature_col3 = ['petal_length']
X3 = df_iris[feature_col3]
#petal_width
feature_col4 = ['petal_width']
X4 = df_iris[feature_col4]



In [13]:
#train data on feature_col1, sepal_length
X1_train, X1_test, y_train, y_test = train_test_split(X1, y, test_size = 0.4,random_state = 6)

In [14]:
#train data on feature_col2, sepal_width
X2_train, X2_test, y_train, y_test = train_test_split(X2, y, test_size = 0.4,random_state = 6)


In [15]:
#train data on feature_col3, petal_length
X3_train, X3_test, y_train, y_test = train_test_split(X3, y, test_size = 0.4,random_state = 6)


In [16]:
#train data on feature_col4, petal_width
X4_train, X4_test, y_train, y_test = train_test_split(X4, y, test_size = 0.4,random_state = 6)


In [17]:
#Instantiate object of KNeighborsClassifier with k = 3
k = 3
knn_kevin1 = KNeighborsClassifier(n_neighbors = k)
knn_kevin2 = KNeighborsClassifier(n_neighbors = k)
knn_kevin3 = KNeighborsClassifier(n_neighbors = k)
knn_kevin4 = KNeighborsClassifier(n_neighbors = k)

In [18]:
#train the model on training set for each feature_col
#sepal_length
knn_kevin1.fit(X1_train, y_train)
#sepal_width
knn_kevin2.fit(X2_train, y_train)
#petal_length
knn_kevin3.fit(X3_train, y_train)
#petal_width
knn_kevin4.fit(X4_train, y_train)

KNeighborsClassifier(n_neighbors=3)

In [19]:
#test model on testing set for each feature_col
#sepal_length
y_predict1 = knn_kevin1.predict(X1_test)
#sepal_width
y_predict2 = knn_kevin2.predict(X2_test)
#petal_length
y_predict3 = knn_kevin3.predict(X3_test)
#petal_width
y_predict4 = knn_kevin4.predict(X4_test)

In [20]:
#find accuracy percentage of each feature individually
#create empty list to store accuracy values
accList = []
#accuracy of sepal_length
accuracy1 = accuracy_score(y_test,y_predict1)
accList.append(accuracy1)

#accuracy of sepal_width
accuracy2 = accuracy_score(y_test,y_predict2)
accList.append(accuracy2)

#accuracy of petal_length
accuracy3 = accuracy_score(y_test,y_predict3)
accList.append(accuracy3)

#accuracy of petal_width
accuracy4 = accuracy_score(y_test,y_predict4)
accList.append(accuracy4)

print(accList)

[0.7166666666666667, 0.5666666666666667, 0.9333333333333333, 0.95]


Which individual provides the best accuracy (the best feature)? 

Feature 4 -> petal_width

Which one is the second best feature?

Feature 3 -> petal_length

In [21]:
#Creating new feature columns based on 2 of 4 features

#sepal_length and sepal_width (features 1 and 2)
feature_col12 = ['sepal_length', 'sepal_width']
X12 = df_iris[feature_col12]

In [22]:
#sepal_length and petal_length (features 1 and 3)
feature_col13 = ['sepal_length', 'petal_length']
X13 = df_iris[feature_col13]

In [23]:
#sepal_length and petal_width (features 1 and 4)
feature_col14 = ['sepal_length' , 'petal_width']
X14 = df_iris[feature_col14]

In [24]:
#sepal_width and petal_length (features 2 and 3)
feature_col23 = ['sepal_width' , 'petal_length']
X23 = df_iris[feature_col23]

In [25]:
#sepal_width and petal_width (features 2 and 4)
feature_col24 = ['sepal_width' , 'petal_width']
X24 = df_iris[feature_col24]

In [26]:
#petal_length and petal_width (features 3 and 4)
feature_col34 = ['petal_length' , 'petal_width']
X34 = df_iris[feature_col34]

In [27]:
#train data on feature_col12, sepal_length and sepal_width
X12_train, X12_test, y_train, y_test = train_test_split(X12, y, test_size = 0.4,random_state = 6)

In [28]:
#train data on feature_col13, sepal_length and petal_length
X13_train, X13_test, y_train, y_test = train_test_split(X13, y, test_size = 0.4,random_state = 6)

In [29]:
#train data on feature_col14, sepal_length and petal_width
X14_train, X14_test, y_train, y_test = train_test_split(X14, y, test_size = 0.4,random_state = 6)

In [30]:
#train data on feature_col23, sepal_width and petal_length
X23_train, X23_test, y_train, y_test = train_test_split(X23, y, test_size = 0.4,random_state = 6)

In [31]:
#train data on feature_col24, sepal_width and petal_width
X24_train, X24_test, y_train, y_test = train_test_split(X24, y, test_size = 0.4,random_state = 6)

In [32]:
#train data on feature_col34, petal_length and petal_width
X34_train, X34_test, y_train, y_test = train_test_split(X34, y, test_size = 0.4,random_state = 6)

In [33]:
#Instantiate object of KNeighborsClassifier with k = 3
k = 3
knn_kevin12 = KNeighborsClassifier(n_neighbors = k)
knn_kevin13 = KNeighborsClassifier(n_neighbors = k)
knn_kevin14 = KNeighborsClassifier(n_neighbors = k)
knn_kevin23 = KNeighborsClassifier(n_neighbors = k)
knn_kevin24 = KNeighborsClassifier(n_neighbors = k)
knn_kevin34 = KNeighborsClassifier(n_neighbors = k)

In [34]:
#train the model on training set for each feature_col pair
#sepal_length12
knn_kevin12.fit(X12_train, y_train)

#sepal_width13
knn_kevin13.fit(X13_train, y_train)

#petal_length14
knn_kevin14.fit(X14_train, y_train)

#petal_width23
knn_kevin23.fit(X23_train, y_train)

#petal_width24
knn_kevin24.fit(X24_train, y_train)

#petal_width34
knn_kevin34.fit(X34_train, y_train)


KNeighborsClassifier(n_neighbors=3)

In [35]:
#test model on testing set for each feature_col pair

#sepal_length and sepal_width
y_predict12 = knn_kevin12.predict(X12_test)

In [36]:
#sepal_length and petal_length
y_predict13 = knn_kevin13.predict(X13_test)

In [37]:
#sepal_length and petal_width
y_predict14 = knn_kevin14.predict(X14_test)

In [38]:
#sepal_width and petal_length
y_predict23 = knn_kevin23.predict(X23_test)

In [39]:
#sepal_width and petal_width
y_predict24 = knn_kevin24.predict(X24_test)

In [40]:
#petal_length and petal_width
y_predict34 = knn_kevin34.predict(X34_test)

In [41]:
#find accuracy percentage of each feature pair
#create empty list to store accuracy values
featList = []

#accuracy of sepal_length and sepal_width
accuracy12 = accuracy_score(y_test,y_predict12)
featList.append(accuracy12)

In [42]:
#accuracy of sepal_length and petal_length
accuracy13 = accuracy_score(y_test,y_predict13)
featList.append(accuracy13)

In [43]:
#accuracy of sepal_length and petal_width
accuracy14 = accuracy_score(y_test,y_predict14)
featList.append(accuracy14)

In [44]:
#accuracy of sepal_width and petal_length
accuracy23 = accuracy_score(y_test,y_predict23)
featList.append(accuracy23)

In [45]:
#accuracy of sepal_width and petal_width
accuracy24 = accuracy_score(y_test,y_predict24)
featList.append(accuracy24)

In [46]:
#accuracy of petal_length and petal_width
accuracy34 = accuracy_score(y_test,y_predict34)
featList.append(accuracy34)

In [47]:
#print list of percentage accuracies
print(featList)

[0.8166666666666667, 0.9833333333333333, 0.95, 0.95, 0.95, 0.9666666666666667]


Which feature pair provides the best accuracy?
Features one (sepal_length) and feature three (petal_length) at a 98.333%

# Part G

The best feature pair from F contains feature one, which is sepal_length, and feature 3, which is petal_length at a 98.333%. In part E, the two best features were feature 4, sepal_width at 95% and feature 3, sepal_length at 93.333%. From this specific example we can conclude that the two best single features are not the best feature pair because the best feature pair, features 1 and 3, yielded a 98% accuracy while features 3 and 4, which are the best single features, yielded a 96.6% accuracy.

# Part H

As I stated above, the best feature pair yielded a 98% accuracy rate while the two best single feautres paired together yielded a 96.6% accuracy rate. Even tho that is a tiny 1.4% difference, 2 other features yeilded a higher rate than the 2 single best, thus elimintaing the ability for the two best single features to have be the best feature pair. There might be error in my code somewhere but from the results shown above my analysis on the situation is correct. 